src/reanimated2/shareables.ts

```diff
@@ -37,16 +37,74 @@ export function registerShareableMapping(
   _shareableCache.set(shareable, shareableRef || _shareableFlag);
 }
 
+// The below object is used as a replacement for objects that cannot be transferred
+// as shareable values. In makeShareableCloneRecursive we detect if an object is of
+// a plain Object.prototype and only allow such objects to be transferred. This lets
+// us avoid all sorts of react internals from leaking into the UI runtime. To make it
+// possible to catch errors when someone actually tries to access such object on the UI
+// runtime, we use the below Proxy object which is instantiated on the UI runtime and
+// throws whenever someone tries to access its fields.
+const INACCESSIBLE_OBJECT = {
+  __init: () => {
+    'worklet';
+    return new Proxy(
+      {},
+      {
+        get: (_: any, prop: string) => {
+          if (prop === '_isReanimatedSharedValue') {
+            // not very happy about this check here, but we need to allow for
+            // "inaccessible" objects to be tested with isSharedValue check
+            // as it is being used in the mappers when extracing inputs recursively.
+            // Apparently we can't check if a key exists there as HostObjects always
+            // return true for such tests, so the only possibility for us is to
+            // actually access that key and see if it is set to true. We therefore
+            // need to allow for this key to be accessed here.
+            return false;
+          }
+          throw new Error(
+            'Trying to access properties of an object which cannot be sent to the UI runtime.'
+          );
+        },
+        set: () => {
+          throw new Error(
+            'Trying to write to an object which cannot be sent to the UI runtime.'
+          );
+        },
+      }
+    );
+  },
+};
+
+const ANALYZE_CYCLIC_OBJECT_AT_DEPTH = 30;
+let CYCLIC_OBJECT_TEST: any;
```

Comment:
I'm not a huge fan of these two names:
1. As for `ANALYZE_CYCLIC_OBJECT_AT_DEPTH`, we analyze all objects, not only the cyclic ones; also it's more of a lower bound for the depth rather than the depth itself
2. As for `CYCLIC_OBJECT_TEST`, this name gives no information on what's exactly being stored